metadata-ingestion/src/datahub/ingestion/source/schema/json_schema.py

```diff
@@ -0,0 +1,364 @@
+        if isinstance(v, AnyHttpUrl):
+            with urllib.request.urlopen(v) as response:
+                schema_dict = json.load(response)
+                if not JsonSchemaTranslator._get_id_from_any_schema(schema_dict):
+                    schema_dict["$id"] = str(v)
+                with tempfile.NamedTemporaryFile(mode="w", delete=False) as tmp_file:
+                    tmp_file.write(json.dumps(schema_dict))
+                    tmp_file.flush()
+                    return tmp_file.name
+        return v
+
+
+class JsonSchemaCheckpointState(GenericCheckpointState):
+    def __init__(self, **data: Any):
+        super().__init__(**data)
+
+
+@platform_name(platform_name="JSON Schemas", id="json-schema")
+@config_class(JsonSchemaSourceConfig)
+@support_status(SupportStatus.INCUBATING)
+@capability(
+    capability_name=SourceCapability.SCHEMA_METADATA,
+    description="Extracts schemas, following references",
+)
+@capability(
+    capability_name=SourceCapability.PLATFORM_INSTANCE,
+    description="Supports platform instance via config",
+)
+@capability(
+    capability_name=SourceCapability.DESCRIPTIONS,
+    description="Extracts descriptions at top level and field level",
+)
+@capability(
+    capability_name=SourceCapability.DELETION_DETECTION,
+    description="With stateful ingestion enabled, will remove entities from DataHub if they are no longer present in the source",
+)
+@capability(
+    capability_name=SourceCapability.TAGS,
+    description="Does not currently support extracting tags",
+    supported=False,
+)
+@capability(
+    capability_name=SourceCapability.OWNERSHIP,
+    description="Does not currently support extracting ownership",
+    supported=False,
+)
+@dataclass
+class JsonSchemaSource(StatefulIngestionSourceBase):
+    """
+    This source extracts metadata from a single JSON Schema or multiple JSON Schemas rooted at a particular path.
+    It performs reference resolution based on the `$ref` keyword.
+
+    Metadata mapping:
+    - Schemas are mapped to Datasets with sub-type Schema
+    - The name of the Schema (Dataset) is inferred from the `$id` property and if that is missing, the file name.
+    - Browse paths are minted based on the path
+    """
+
+    @staticmethod
+    def _pull_up_reference(schema_dict):
+        """Transforms a schema that is of the form
+        {
+         "$ref": "#definitions/MyDefinition",
+         "definitions": {
+            "MyDefinition": {
+                ... definition
+            },
+            "OtherDefinition": {
+                ... other definition
+            }
+         }
+        }
+        to
+        {
+            ... definition
+
+            "definitions": {
+            "OtherDefinition": {
+                ... other definition
+            }
+         }
+        """
+
+        if "$ref" in schema_dict:
+            reference = schema_dict["$ref"]
+            if reference.startswith("#/"):
+                parts = reference[2:].split("/")
```

Comment:
nitpick: would like to see this tedious logic pulled into smaller well named functions

Reply:
this is already a function